File: r_code/data_prep/data_gen_functions.R
#' Creating a simulated count matrix
#' 
#' @param output_path Path to data folder, where count matrix is saved (mandatory)
#' @param number_of_genes Number of genes to have in the matrix (default is 1000)
#' @param number_of_samples Number of samples to have in the matrix (default is 10)
#' @param seed Random seed for reproducibility (default is no seed)
#' @returns count_matrix
#' @examples
#' generate_counts(
#'  output_path = "output/data/",
#'  number_of_genes = 1000,
#'  number_of_samples = 10,
#'  seed = 567)
generate_counts <- function(
    output_path,
    number_of_genes,
    number_of_samples,
    seed = NA) {
    
    if (!is.na(seed)) {
        set.seed(seed)
    }

    count_matrix <- matrix(
        data = round(runif(number_of_genes * number_of_samples, min = 0, max = 100)),
        nrow = number_of_genes)
    colnames(count_matrix) <- paste0("Sample", 1:number_of_samples)
    rownames(count_matrix) <- paste0("Gene", 1:number_of_genes)

    save(
        count_matrix,
        file = file.path(output_path, "count_matrix.RData"))
    write.csv(
        count_matrix,
        file=file.path(output_path, "count_matrix.csv"))
    return(count_matrix)
}


#' Creating a simulated metadata associated to count matrix
#' 
#' @param output_path Path to data folder, where metadata is saved (mandatory)
#' @param number_of_samples Number of samples to have in the matrix (default is 10)
#' @param condition_vector Names of different conditions (default is "Treatment" and "Control")
#' @returns list(counts = count_matrix, metadata = metadata)
#' @examples
#' generate_metadata(
#'  output_path = DATA_DIR,
#'  number_of_samples = 10,
#'  condition_vector = c("Control", "Treatment"))
generate_metadata <- function(
    output_path,
    count_matrix,
    number_of_samples,
    condition_vector = c("Control", "Treatment")){
    metadata <- data.frame(
        Sample = colnames(count_matrix),
        Condition = sample(
            condition_vector, 
            size = number_of_samples,
            replace = TRUE),
        stringsAsFactors = FALSE)
    
    data.table::fwrite(
        metadata,
        file = file.path(output_path, "metadata.tsv"),
        sep = "\t",
        col.names = TRUE)

    return(metadata)
}

#' Creating a simulated count matrix and associated metadata
#' 
#' @param output_path Path to data folder, where count matrix and metadata is saved (mandatory)
#' @param number_of_genes Number of genes to have in the matrix (default is 1000)
#' @param number_of_samples Number of samples to have in the matrix (default is 10)
#' @param seed Random seed for reproducibility (default is no seed)
#' @param condition_vector Names of different conditions (default is "Treatment" and "Control")
#' @returns list(counts = count_matrix, metadata = metadata)
#' @examples
#' generate_counts_with_metadata(
#'  output_path = DATA_DIR,
#'  number_of_genes = 1000,
#'  number_of_samples = 10,
#'  seed = 567,
#'  condition_vector = c("Control", "Treatment"))
generate_counts_with_metadata <- function(
    output_path,
    number_of_genes = 1000,
    number_of_samples = 10,
    seed = NA,
    condition_vector = c("Control", "Treatment")) {

    ## Generating counts:
    count_matrix <- generate_counts(
        output_path = output_path,
        number_of_genes = number_of_genes,
        number_of_samples = number_of_samples,
        seed = seed)

    ## Generating metadata:
    metadata <- generate_metadata(
        output_path = output_path,
        count_matrix = count_matrix,
        number_of_samples = number_of_samples,
        condition_vector = condition_vector)

    return(list(
        counts = count_matrix,
        metadata = metadata))

}

File: python_code/data_prep/data_gen_functions.py
import numpy as np
import pandas as pd

from pathlib import Path
from typing import Optional, Tuple

def generate_counts(
    output_path: Path,
    number_of_genes: int = 1000,
    number_of_samples: int = 10,
    rng: Optional[np.random.Generator] = None,
    seed: Optional[int] = None,
) -> pd.DataFrame:

    if rng is None:
        rng = np.random.default_rng(seed)

    counts = rng.integers(0, 100, size=(number_of_genes, number_of_samples), 
        endpoint=True)
    colnames = [f'Sample{i}' for i in range(1, number_of_samples+1)]
    rownames = [f'Gene{i}' for i in range(1, number_of_genes+1)]
    count_df = pd.DataFrame(counts, index=rownames, columns=colnames)

    count_df.to_pickle(output_path)

    return count_df


def generate_metadata(
    metadata_output_path: Path,
    condition_vector = ['Control', 'Treatment'],
    number_of_samples: int = 10,
    rng: Optional[np.random.Generator] = None,
    seed: Optional[int] = None,
) -> pd.DataFrame:
    
    if rng is None:
        rng = np.random.default_rng(seed)
    
    conditions = rng.choice(condition_vector, number_of_samples)
    colnames = [f'Sample{i}' for i in range(1, number_of_samples+1)]
    metadata = pd.DataFrame(conditions, index=colnames, columns=['Condition'])

    metadata.to_csv(metadata_output_path, sep='\t')

    return metadata


def generate_counts_with_metadata(
    counts_output_path: Path,
    metadata_output_path: Path,
    number_of_genes: int = 1000,
    number_of_samples: int = 10,
    condition_vector = ['Control', 'Treatment'],
    rng: Optional[np.random.Generator] = None,
    seed: Optional[int] = None,
) -> Tuple[pd.DataFrame, pd.DataFrame]:

    if rng is None:
        rng = np.random.default_rng(seed)
    
    count_df = generate_counts(counts_output_path, number_of_genes, 
        number_of_samples, rng)
    metadata = generate_metadata(metadata_output_path, condition_vector, 
        number_of_samples, rng)
    
    return (count_df, metadata)